usecases/telemetry/telemetry.go

```diff
@@ -0,0 +1,211 @@
+	consumerURL       string
+	pushInterval      time.Duration
+}
+
+// New creates a new Telemeter instance
+func New(nodesStatusGetter nodesStatusGetter, modulesProvider modulesProvider,
+	logger logrus.FieldLogger,
+) *Telemeter {
+	tel := &Telemeter{
+		machineID:         strfmt.UUID(uuid.NewString()),
+		nodesStatusGetter: nodesStatusGetter,
+		modulesProvider:   modulesProvider,
+		logger:            logger,
+		shutdown:          make(chan struct{}),
+		consumerURL:       defaultConsumerURL,
+		pushInterval:      defaultPushInterval,
+	}
+	return tel
+}
+
+// Start begins telemetry for the node
+func (tel *Telemeter) Start(ctx context.Context) error {
+	payload, err := tel.push(ctx, PayloadType.Init)
+	if err != nil {
+		tel.failedToStart = true
+		return fmt.Errorf("push: %w", err)
+	}
+	go func() {
+		t := time.NewTicker(tel.pushInterval)
+		for {
+			select {
+			case <-tel.shutdown:
+				return
+			case <-t.C:
+				payload, err = tel.push(ctx, PayloadType.Update)
+				if err != nil {
+					tel.logger.
+						WithField("action", "telemetry_push").
+						WithField("payload", fmt.Sprintf("%+v", payload)).
+						WithField("retry_at", time.Now().Add(tel.pushInterval).Format(time.RFC3339)).
+						Error(err.Error())
+					return
+				}
+				tel.logger.
+					WithField("action", "telemetry_push").
+					WithField("payload", fmt.Sprintf("%+v", payload)).
+					Info("telemetry update")
+			}
+		}
+	}()
+
+	tel.logger.
+		WithField("action", "telemetry_push").
+		WithField("payload", fmt.Sprintf("%+v", payload)).
+		Info("telemetry started")
+	return nil
+}
+
+// Stop shuts down the telemeter
+func (tel *Telemeter) Stop(ctx context.Context) error {
+	if tel.failedToStart {
+		return nil
+	}
+
+	waitForShutdown := make(chan struct{})
+	go func() {
+		tel.shutdown <- struct{}{}
+		waitForShutdown <- struct{}{}
+	}()
+
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("shutdown telemetry: %w", ctx.Err())
+	case <-waitForShutdown:
+		payload, err := tel.push(ctx, PayloadType.Terminate)
+		if err != nil {
+			tel.logger.
+				WithField("action", "telemetry_push").
+				WithField("payload", fmt.Sprintf("%+v", payload)).
+				Error(err.Error())
+			return err
+		}
+		tel.logger.
+			WithField("action", "telemetry_push").
+			WithField("payload", fmt.Sprintf("%+v", payload)).
+			Info("telemetry terminated")
+		return nil
+	}
+}
+
+// push sends telemetry data to the consumer url
+func (tel *Telemeter) push(ctx context.Context, payloadType string) (*Payload, error) {
+	payload, err := tel.buildPayload(ctx, payloadType)
+	if err != nil {
+		return nil, fmt.Errorf("build payload: %w", err)
+	}
+
+	b, err := json.Marshal(payload)
+	if err != nil {
+		return nil, fmt.Errorf("marshal payload: %w", err)
+	}
+
+	resp, err := http.Post(tel.consumerURL, "application/json", bytes.NewReader(b))
+	if err != nil {
+		return nil, fmt.Errorf("failed to send request: %w", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		defer resp.Body.Close()
```

Comment:
shall we close it no matter the status code to avoid leaks, or we have a specific reason  ?

Reply:
My assumption was that the resp.Body is only `Read()` if status != 200, so no need to close otherwise, since the `io.ReadCloser` is never opened.

However, your question let me to investigate further, and the [docs](https://tip.golang.org/pkg/net/http/?m=old#Response) suggest that the body should always be closed:

```
    // The http Client and Transport guarantee that Body is always
    // non-nil, even on responses without a body or responses with
    // a zero-length body. It is the caller's responsibility to
    // close Body. The default HTTP client's Transport may not
    // reuse HTTP/1.x "keep-alive" TCP connections if the Body is
    // not read to completion and closed.
```

Although I can't tell with 100% certainty that not closing an unread response body will lead to leaks (maybe you know better 😉) I can for sure add the close to all conditions. It won't hurt anything